File: flutter/packages/hatake_core/lib/src/format/form_normalizer.dart
import '../definition/form_definition.dart';
import '../repository/repository.dart';
import 'converter_registry.dart';

/// Applies each field's `normalize` converter chain to a record — run before
/// validation / persistence so input is cleaned consistently on any tier.
class FormNormalizer {
  final ConverterRegistry registry;

  FormNormalizer([ConverterRegistry? registry])
      : registry = registry ?? ConverterRegistry();

  /// Returns a copy of [record] with each field's `normalize` converters
  /// applied (in order). Fields absent from the record are left untouched.
  DataRecord normalize(FormDefinition form, DataRecord record) {
    final result = {...record};
    for (final field in form.fields) {
      if (field.normalize.isEmpty) continue;
      if (!result.containsKey(field.field)) continue;
      result[field.field] =
          registry.convertAll(field.normalize, result[field.field]);
    }
    return result;
  }
}

File: java/src/main/java/io/hatake/core/FormNormalizer.java
package io.hatake.core;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Applies each field's {@code normalize} converter chain to a record — run
 * before validation / persistence so input is cleaned consistently.
 */
public final class FormNormalizer {

    private final ConverterRegistry registry;

    public FormNormalizer() {
        this(new ConverterRegistry());
    }

    public FormNormalizer(ConverterRegistry registry) {
        this.registry = registry;
    }

    public Map<String, Object> normalize(FormDefinition form, Map<String, Object> record) {
        Map<String, Object> out = new LinkedHashMap<>(record);
        for (FieldDefinition f : form.fields()) {
            if (f.normalize().isEmpty() || !out.containsKey(f.field())) {
                continue;
            }
            out.put(f.field(), registry.convertAll(f.normalize(), out.get(f.field())));
        }
        return out;
    }
}
